homeassistant/components/modbus/__init__.py

```diff
@@ -88,163 +97,219 @@
+            self._config_stopbits = client_config[CONF_STOPBITS]
+            self._config_bytesize = client_config[CONF_BYTESIZE]
+            self._config_parity = client_config[CONF_PARITY]
+        else:
+            # network configuration
+            self._config_host = client_config[CONF_HOST]
 
     @property
     def name(self):
         """Return the name of this hub."""
-        return self._name
-
-    def close(self):
-        """Disconnect client."""
-        with self._lock:
-            self._client.close()
+        return self._config_name
+
+    async def _connect_delay(self):
+        if self._config_delay > 0:
+            await asyncio.sleep(self._config_delay)
+            self._config_delay = 0
+
+    def setup(self):
+        """Set up pymodbus client."""
+        # pylint: disable = E0633
+        # Client* do deliver loop, client as result but
+        # pylint does not accept that fact
+
+        _LOGGER.debug("doing setup")
+        if self._config_type == "serial":
+            _, self._client = ClientSerial(
+                schedulers.ASYNC_IO,
+                method=self._config_method,
+                port=self._config_port,
+                baudrate=self._config_baudrate,
+                stopbits=self._config_stopbits,
+                bytesize=self._config_bytesize,
+                parity=self._config_parity,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "rtuovertcp":
+            _, self._client = ClientTCP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                framer=ModbusRtuFramer,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "tcp":
+            _, self._client = ClientTCP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        elif self._config_type == "udp":
+            _, self._client = ClientUDP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        else:
+            assert False
 
-    def connect(self):
-        """Connect client."""
-        with self._lock:
-            self._client.connect()
+    async def _read(self, unit, address, count, func):
+        """Read generic with error handling."""
+        await self._connect_delay()
+        async with self._lock:
+            kwargs = {"unit": unit} if unit else {}
+            result = await func(address, count, **kwargs)
+            if isinstance(result, (ModbusException, ExceptionResponse)):
```

Comment:
It would be nice if the client func raised errors instead of returning them. I guess this is the library responsibility.